compiler/src/iree/compiler/Dialect/HAL/IR/HALOps.td

```diff
@@ -1863,6 +1845,57 @@ def HAL_DeviceQueueFlushOp : HAL_Op<"device.queue.flush"> {
 
 } // OpGroupDeviceOps
 
+//===----------------------------------------------------------------------===//
+// !hal.device management
+//===----------------------------------------------------------------------===//
+
+def OpGroupDeviceManagementOps : OpDocGroup {
+  let summary = "Device management ops";
+  let description = "Device availability and selection support.";
+}
+
+let opDocGroup = OpGroupDeviceManagementOps in {
+
+def HAL_DevicesCountOp : HAL_PureOp<"devices.count", [
+  DeclareOpInterfaceMethods<OpAsmOpInterface, ["getAsmResultNames"]>,
+]> {
+  let summary = [{returns the number of available devices}];
+  let description = [{
+    Returns the total number of available devices registered at runtime.
+  }];
+
+  let results = (outs
+    Index:$result
+  );
+
+  let assemblyFormat = [{
+    attr-dict `:` type($result)
+  }];
+}
+
+def HAL_DevicesGetOp : HAL_PureOp<"devices.get", [
+  DeclareOpInterfaceMethods<OpAsmOpInterface, ["getAsmResultNames"]>,
+]> {
+  let summary = [{returns the device with the given index}];
+  let description = [{
+    Returns the device with the given index in the [0, hal.devices.count) range.
+    Devices may be lazily initialized upon first use.
+  }];
```

Comment:
> The plan is to have a `hal.devices.lookup` op that is hoisted into an initializer to enumerate devices and store the selected ones in globals. The enumeration code will emit a loop over `hal.devices.count` and use `hal.devices.get` and the various device query methods to match the requested devices.

So will the HAL dialect provide ways to see details about each device, such as supported executable formats and this info? https://github.com/openxla/iree/blob/3654fd683c1cbb5df76023dd0880ca164fba55c8/runtime/src/iree/hal/device.h#L72-L80

`hal.device.query` (reference: https://iree.dev/reference/mlir-dialects/HAL/#haldevicequery-haldevicequeryop) doesn't look like enough on its own.

If the compiler doesn't actually need that information, what will it do instead? I imagine what the compiler needs ahead of time is pretty different from what a programmer would need. If I were writing code against this, I would want to loop over all devices and find "the discrete GPU device", for example.

Reply:
device query already exposes the ability to check for executable format support and I'll be adding additional queries as needed

Reply:
Oh, great :D.

I'd forgotten that `hal.executable.format` was a "well-known key" (should have read the page I linked...). I was thinking that the keys would be device-specific and common queries would have dedicated methods.